File: frontend/src/types/auth10.ts
/** Domain types for auth / AuthRecord10 */
export type AuthRecord10Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface AuthRecord10 {
  id: string;
  externalCode: string;
  displayName: string;
  status: AuthRecord10Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface AuthRecord10CreateRequest {
  externalCode: string;
  displayName: string;
  status: AuthRecord10Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface AuthRecord10ListResponse {
  items: AuthRecord10[];
  total?: number;
}

export interface AuthRecord10StatsResponse {
  activeCount: number;
}

export function isAuthRecord10Active(rec: AuthRecord10): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatAuthRecord10Label(rec: AuthRecord10): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareAuthRecord10ByPriority(a: AuthRecord10, b: AuthRecord10): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}

File: backend/internal/service/auth_10.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// AuthRecord10Service encapsulates business rules for AuthRecord10.
type AuthRecord10Service struct {
	repo *repository.AuthRecord10Repository
}

// NewAuthRecord10Service wires the service.
func NewAuthRecord10Service(repo *repository.AuthRecord10Repository) *AuthRecord10Service {
	return &AuthRecord10Service{repo: repo}
}

// Create validates and persists a new AuthRecord10.
func (s *AuthRecord10Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.AuthRecord10, error) {
	now := time.Now().UTC()
	rec := &models.AuthRecord10{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a AuthRecord10 by id.
func (s *AuthRecord10Service) Get(ctx context.Context, id string) (*models.AuthRecord10, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped AuthRecord10 pages.
func (s *AuthRecord10Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.AuthRecord10, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing AuthRecord10.
func (s *AuthRecord10Service) Update(ctx context.Context, actor string, rec *models.AuthRecord10) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.UpdatedBy = actor
	rec.UpdatedAt = time.Now().UTC()
	// repository Update applies optimistic locking and increments version
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a AuthRecord10.
func (s *AuthRecord10Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds AuthRecord10 by external code prefix.
func (s *AuthRecord10Service) Search(ctx context.Context, facilityID, prefix string) ([]models.AuthRecord10, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *AuthRecord10Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}
